evm_loader/program/src/instruction/transaction_step.rs

```diff
@@ -48,17 +44,18 @@ pub fn do_continue<'a>(
     }
 
     let account_storage = ProgramAccountStorage::new(accounts)?;
-    let mut persistent_state  = PersistentState::restore(&account_storage);
+    let mut persistent_state = PersistentState::restore(&account_storage);
 
     let (result, steps_executed, _) = {
         match persistent_state.backend.exit_status() {
             Some(status) => (status.clone(), 0_u64, None),
             None => {
                 // TODO: BORROWCHECKER CURSES HERE (FOR A REASON).
                 // I NEED TO FIGURE OUT THE WAY TO AVOID IT WITHOUT MAJOR CHANGES.
+                let persistent_state = &mut *persistent_state;
```

Comment:
`ManuallyDrop` implements `DerefMut` trait, so each access like `persistent_state.backend` and `persistent_state.root_evm` calls `DerefMut` implementation and borrows `&mut self`.

By forcing the `persistent_state.backend` and `persistent_state.root_evm` borrows on a `&mut ManuallyDrop` reference (created by `&mut *persistent_state`), the compiler can use the [`Splitting Borrows`](https://doc.rust-lang.org/nomicon/borrow-splitting.html) feature again.

The error you encountered is a common one for all smart pointer types implementing `BorrowMut` trait: https://github.com/rust-lang/rust/issues/72297

I suggest reading these two resources, they helped me understand Rust borrow checker and lifetimes better:
- https://quinedot.github.io/rust-learning/index.html
- https://github.com/pretzelhammer/rust-blog/blob/master/posts/common-rust-lifetime-misconceptions.md